File: frontend/appflowy_tauri/src/appflowy_app/stores/effects/database/field/field_listener.ts
import { Err, Ok, Result } from 'ts-results';
import { DatabaseNotification } from '../../../../../services/backend';
import { DatabaseFieldChangesetPB } from '../../../../../services/backend/models/flowy-database/field_entities';
import { FlowyError } from '../../../../../services/backend/models/flowy-error';
import { ChangeNotifier } from '../../../../utils/change_notifier';
import { DatabaseNotificationListener } from '../../notifications/listener';

type UpdateFieldNotifiedValue = Result<DatabaseFieldChangesetPB, FlowyError>;
export type DatabaseNotificationCallback = (value: UpdateFieldNotifiedValue) => void;

export class DatabaseFieldListener {
  _notifier?: ChangeNotifier<UpdateFieldNotifiedValue>;
  _listener?: DatabaseNotificationListener;

  constructor(public readonly databaseId: string) {}

  start = (callbacks: { onFieldsChanged: DatabaseNotificationCallback }) => {
    this._notifier = new ChangeNotifier();
    this._notifier?.observer.subscribe(callbacks.onFieldsChanged);

    this._listener = new DatabaseNotificationListener({
      viewId: this.databaseId,
      parserHandler: (notification, payload) => {
        switch (notification) {
          case DatabaseNotification.DidUpdateFields:
            this._notifier?.notify(Ok(DatabaseFieldChangesetPB.deserializeBinary(payload)));
            return;
          default:
            break;
        }
      },
      onError: (error) => this._notifier?.notify(Err(error)),
    });
    return undefined;
  };

  stop = async () => {
    this._notifier?.unsubscribe();
    await this._listener?.stop();
  };
}

File: frontend/appflowy_flutter/lib/plugins/database/application/field/field_listener.dart
import 'package:appflowy/core/notification/grid_notification.dart';
import 'package:appflowy_backend/log.dart';
import 'package:dartz/dartz.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/notification.pb.dart';
import 'package:flowy_infra/notifier.dart';
import 'dart:async';
import 'dart:typed_data';
import 'package:appflowy_backend/protobuf/flowy-database2/field_entities.pb.dart';

typedef UpdateFieldNotifiedValue = FieldPB;

class SingleFieldListener {
  final String fieldId;
  void Function(UpdateFieldNotifiedValue)? _updateFieldNotifier;
  DatabaseNotificationListener? _listener;

  SingleFieldListener({required this.fieldId});

  void start({
    required void Function(UpdateFieldNotifiedValue) onFieldChanged,
  }) {
    _updateFieldNotifier = onFieldChanged;
    _listener = DatabaseNotificationListener(
      objectId: fieldId,
      handler: _handler,
    );
  }

  void _handler(
    DatabaseNotification ty,
    Either<Uint8List, FlowyError> result,
  ) {
    switch (ty) {
      case DatabaseNotification.DidUpdateField:
        result.fold(
          (payload) => _updateFieldNotifier?.call(FieldPB.fromBuffer(payload)),
          (error) => Log.error(error),
        );
        break;
      default:
        break;
    }
  }

  Future<void> stop() async {
    await _listener?.stop();
    _updateFieldNotifier = null;
  }
}

typedef UpdateFieldsNotifiedValue
    = Either<DatabaseFieldChangesetPB, FlowyError>;

class FieldsListener {
  final String viewId;
  PublishNotifier<UpdateFieldsNotifiedValue>? updateFieldsNotifier =
      PublishNotifier();
  DatabaseNotificationListener? _listener;
  FieldsListener({required this.viewId});

  void start({
    required void Function(UpdateFieldsNotifiedValue) onFieldsChanged,
  }) {
    updateFieldsNotifier?.addPublishListener(onFieldsChanged);
    _listener = DatabaseNotificationListener(
      objectId: viewId,
      handler: _handler,
    );
  }

  void _handler(DatabaseNotification ty, Either<Uint8List, FlowyError> result) {
    switch (ty) {
      case DatabaseNotification.DidUpdateFields:
        result.fold(
          (payload) => updateFieldsNotifier?.value =
              left(DatabaseFieldChangesetPB.fromBuffer(payload)),
          (error) => updateFieldsNotifier?.value = right(error),
        );
        break;
      default:
        break;
    }
  }

  Future<void> stop() async {
    await _listener?.stop();
    updateFieldsNotifier?.dispose();
    updateFieldsNotifier = null;
  }
}
